clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +375,324 @@ impl FunctionType {
+        ) => {
+            let contract_to_check = match db.load_contract(&contract_identifier) {
+                Some(contract) => {
+                    runtime_cost(
+                        ClarityCostFunction::AnalysisFetchContractEntry,
+                        tracker,
+                        contract_analysis_size(&contract)?,
+                    )?;
+                    contract
+                }
+                None => {
+                    runtime_cost(ClarityCostFunction::AnalysisFetchContractEntry, tracker, 1)?;
+                    return Err(CheckErrors::NoSuchContract(contract_identifier.to_string()).into());
+                }
+            };
+            let expected_trait = lookup_trait(
+                db,
+                contract_context,
+                expected_trait_id,
+                clarity_version,
+                tracker,
+            )?;
+            contract_to_check.check_trait_compliance(expected_trait_id, &expected_trait)?;
+        }
+        (
+            TypeSignature::ListUnionType(types),
+            TypeSignature::CallableType(CallableSubtype::Trait(_)),
+        ) => {
+            // Verify that all types in the union implement this trait
+            for subtype in types {
+                inner_type_check_type(
+                    db,
+                    contract_context,
+                    &TypeSignature::CallableType(subtype.clone()),
+                    expected_type,
+                    clarity_version,
+                    depth + 1,
+                    tracker,
+                )?;
+            }
+        }
+        (TypeSignature::NoType, _) => (),
+        (_, _) => {
+            if !expected_type.admits_type(&actual_type) {
+                return Err(
+                    CheckErrors::TypeError(expected_type.clone(), actual_type.clone()).into(),
+                );
+            }
+        }
+    }
+    Ok(expected_type.clone())
+}
+
+fn lookup_trait<T: CostTracker>(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    trait_id: &TraitIdentifier,
+    clarity_version: ClarityVersion,
+    tracker: &mut T,
+) -> CheckResult<BTreeMap<ClarityName, FunctionSignature>> {
+    if let Some(contract_context) = contract_context {
+        // If the trait is from this contract, then it must be in the context or it doesn't exist.
+        if contract_context.is_contract(&trait_id.contract_identifier) {
+            return Ok(contract_context
+                .get_trait(&trait_id, clarity_version)
+                .ok_or(CheckErrors::NoSuchTrait(
+                    trait_id.contract_identifier.to_string(),
+                    trait_id.name.to_string(),
+                ))?
+                .clone());
+        }
+        if clarity_version >= ClarityVersion::Clarity2 {
+            if let Some(trait_sig) = contract_context.get_trait(trait_id, clarity_version) {
+                return Ok(trait_sig.clone());
+            }
```

Comment:
Should this also short-circuit if the trait isn't found, like:

```rust
            if let Some(trait_sig) = contract_context.get_trait(trait_id, clarity_version) {
                return Ok(trait_sig.clone());
            } else {
                return Err(CheckErrors::NoSuchTrait(..));
           }
```

Reply:
No, I think we want to continue to do the database lookup in that case.

Reply:
I'm not clear on why?

Reply:
If it's to assess runtime costs, then we should document this as a Clarity1-compatibility behavior.

Reply:
In fact, given the multiple instances of gating here, it might make sense to factor this into two functions -- `clarity2_lookup_trait` and `clarity1_lookup_trait`, where no gating happens within their bodies.  The caller would check `clarity_version` and call the right one.

Reply:
This could be looking up a trait which we have not seen yet, so we'd need to go to the database to retrieve it. For example, if a `contract-call?`takes a trait parameter, we might not have seen that trait yet, since it would not have required a `use-trait`.

Reply:
I've refactored things so that we only use `lookup_trait` for Clarity2. Renamed the function to make that clear and removed the clarity version parameter.

Reply:
ccf363a64